gateway-service/src/main/java/org/zowe/apiml/gateway/zaas/ZaasController.java

```diff
@@ -12,108 +12,94 @@
-                .body(zaasTokenResponse);
-
-        } catch (Exception e) {
-            ApiMessageView messageView = messageService.createMessage("org.zowe.apiml.zaas.zosmf.noZosmfTokenReceived", e.getMessage()).mapToView();
-            return ResponseEntity
-                .status(HttpStatus.SERVICE_UNAVAILABLE)
-                .body(messageView);
-        }
+    public ResponseEntity<ZaasTokenResponse> getZosmfToken(@RequestAttribute(AUTH_SOURCE_ATTR) AuthSource authSource,
+                                                           @RequestAttribute(AUTH_SOURCE_PARSED_ATTR) AuthSource.Parsed authSourceParsed) throws ServiceNotFoundException {
+
+        ZaasTokenResponse zaasTokenResponse = zosmfService.exchangeAuthenticationForZosmfToken(authSource.getRawSource().toString(), authSourceParsed);
+
+        return ResponseEntity
+            .status(HttpStatus.OK)
+            .body(zaasTokenResponse);
     }
 
 
     @PostMapping(path = "zoweJwt", produces = MediaType.APPLICATION_JSON_VALUE)
     @Operation(summary = "Provides zoweJwt for authenticated user.")
-    public ResponseEntity<Object> getZoweJwt(@RequestAttribute(AUTH_SOURCE_ATTR) AuthSource authSource) {
-        try {
-            String token = authSourceService.getJWT(authSource);
-
-            return ResponseEntity
-                .status(HttpStatus.OK)
-                .body(new ZaasTokenResponse(COOKIE_AUTH_NAME, token));
-
-        } catch (Exception e) {
-            ApiMessageView messageView = messageService.createMessage("org.zowe.apiml.zaas.zoweJwt.noToken", e.getMessage()).mapToView();
-            return ResponseEntity
-                .status(HttpStatus.INTERNAL_SERVER_ERROR)
-                .body(messageView);
+    public ResponseEntity<ZaasTokenResponse> getZoweJwt(@RequestAttribute(AUTH_SOURCE_ATTR) AuthSource authSource) {
+
+        String token = authSourceService.getJWT(authSource);
+
+        return ResponseEntity
+            .status(HttpStatus.OK)
+            .body(new ZaasTokenResponse(COOKIE_AUTH_NAME, token));
+    }
+
+    @PostMapping(path = "safIdt", produces = MediaType.APPLICATION_JSON_VALUE)
+    @Operation(summary = "Provides SAF Identity Token for authenticated user.")
+    public ResponseEntity<ZaasTokenResponse> getSafIdToken(@RequestBody TicketRequest ticketRequest, @RequestAttribute(AUTH_SOURCE_PARSED_ATTR) AuthSource.Parsed authSourceParsed)
+        throws IRRPassTicketGenerationException, ApplicationNameNotFoundException {
+
+        final String applicationName = ticketRequest.getApplicationName();
+        if (StringUtils.isBlank(applicationName)) {
+            throw new ApplicationNameNotFoundException("ApplicationName not provided.");
         }
+
+        String safIdToken = tokenCreationService.createSafIdTokenWithoutCredentials(authSourceParsed.getUserId(), applicationName);
+        return ResponseEntity
+            .status(HttpStatus.OK)
+            .body(new ZaasTokenResponse("", safIdToken));
```

Comment:
```suggestion
            .body(new ZaasTokenResponse(null, safIdToken));
```
Empty string or null?

Reply:
Empty string is better option in my opinion.